vcd/resource_vcd_vapp_vm.go

```diff
@@ -1172,6 +1241,59 @@ func createVmEmpty(d *schema.ResourceData, meta interface{}, vmType typeOfVm) (*
 		return nil, fmt.Errorf("`hardware_version` is required when creating empty VM")
 	}
 
+	// supportsFirmware is a helper variable to not repeat the same API checking calls
+	supportsFirmware := vcdClient.Client.APIVCDMaxVersionIs(">=37.1")
+
+	// There are four scenarios:
+	// 1. If firmware is set and VCD is 10.4.1 or newer, we keep the set value
+	// 2. If firmware is unset and VCD is 10.4.1 or newer, we set it to BIOS as it is required by the API
+	// 3. If firmware is set and VCD is 10.4.0 or older, we return an error
+	// 4. If firmware is unset and VCD is 10.4.0 or older, we pass
+	var firmware interface{}
+	if firmware, ok = d.GetOk("firmware"); ok {
+		if !supportsFirmware {
+			return nil, fmt.Errorf("firmware is supported in VCD 10.4.1+")
+		}
+	} else {
+		if supportsFirmware {
+			firmware = "bios"
+		}
+	}
+
+	// Additional checks if the hardware version and OS support new Firmware and BootOptions fields (37.1+)
+	if supportsFirmware {
+		hwVersion, err := vdc.GetHardwareVersion(hardWareVersion.(string))
+		if err != nil {
+			return nil, fmt.Errorf("error getting hardware version: %s", err)
+		}
+
+		os, err := vdc.FindOsFromId(hwVersion, osType.(string))
+		if err != nil {
+			return nil, fmt.Errorf("error finding given OS type: %s", err)
+		}
+
+		if !contains(os.SupportedFirmware, firmware.(string)) {
+			return nil, fmt.Errorf("provided `os_type` doesn't support provided firmware")
+		}
+	}
```

Comment:
If you can remind me - these errors directly from API were ugly, right?

Reply:
This one was alright, actually better than mine probably : `Firmware efi not supported by VM TestVM with guest operating system sles10_64Guest at hardware version HW11.` It's mostly for `Update` so we don't do anything to the VM like PowerOff or partial update but I guess I could improve the error message.